agent: Make intent of signaling channels clear and optimize memory

aditighag opened the pull request:
When channels are merely used for signaling purposes, use an empty
struct as the channel type since the value of the channel is never read.
Also, this can help with memory optimizations as the empty struct occupies 0 bytes of storage.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

christarazi reviewed (approved):
:rocket:

tklauser commented on pkg/health/server/prober.go:
```diff
@@ -43,8 +43,8 @@ type prober struct {
-	proberExited chan bool
+	stop         chan struct{}
+	proberExited chan struct{}
 	done         chan bool
```
I think the `done` channel could be changed to `chan struct{}` as well, though it requires changing `(*prober).Done()`.

Upon closer inspection this func even looks unused, so it doesn't look like anything is checking the channel being closed, so maybe we don't need the `done` channel after all.

aditighag replied:
Thanks...Done!

tklauser reviewed (approved):
Nice change. One comment inline which could be addressed in a follow-up PR.

gandro reviewed (approved):
I agree with https://github.com/cilium/cilium/pull/14075#discussion_r526698797

We can probably just remove the `prober.Done()` function and `prober.done` channel in `pkg/health` to futher clean up the code.

The existing code waits for `prober.Stop()` to return and does not leak the `prober`, so there is really no use for `prober.Done()` as far as I can tell.

Not blocking the PR on this however.

aditighag committed:
cleanup: Removed the unused function and channel

Suggested-by: Tobias Klauser <tklauser@distanz.ch>
Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
agent: Make intent of signaling channels clear and optimize memory

When channels are used merely for signalling purposes, use an empty
struct as the channel type since the value of the channel is never read.
Also, this can help with memory optimizations since the empty struct occupies 0 bytes of storage.

Signed-off-by: Aditi Ghag <aditi@cilium.io>
